### ISSUE TITLE
Compatibility with OpenDrive and CommonDrive format

### ISSUE THREAD

**Issue by samuruph** [metadriverse/scenarionet#54]
Hello,
Do you know whether ScenarioNet format is compatible in some way with OpenDrive and/or CommonDrive data format?

**Comment by QuanyiLi** [comment]
Yes. Just extract lane center lines/polygons and lane lines from these map format. Then the map can be loaded into the simulator. We don't implement such a converter, so you need to do it on your own. We can help with it.

**Comment by samueleruffino99** [comment]
Do you have any reference to both data format used ? Because I would like to measure criticality in nuScenes scenes by [CommonRoad-CriMe](https://commonroad.in.tum.de/tools/commonroad-crime), metrics [here](https://mediatum.ub.tum.de/doc/1706306/1706306.pdf). Maybe it would be easier to just adapt these metrics to your data format instead of convert from nuScenes to ScenarioNet and then to CommonRoad format.

**Comment by QuanyiLi** [comment]
An example of ScenarioNet data format is like [this](https://metadrive-simulator.readthedocs.io/en/latest/scenario_description.html#what-is-the-scenario-description). Basically, it is a dictionary where you have vehicle trajectories/velocities/headings, and map structures represented by polyline and polygons. And everything is explicit and human-readable. The nuScenes data has no explicit format, you have to access everything in the scenario by calling the original APIs which take their database file as input. For OpenDrive, it is a map format stored `.xml` file which also recorded in lane center lines and lane lines, which can be easily converted from ScenarioNet format (I guess)

Regarding your question, the solution in my mind is using ScenarioNet to extract explicit scenario data from the nuScenes dataset. And then making a data loader for CommonRoad-CriMe, allowing it to parse trajectories and map structures from the dictionary-like ScenarioNet format.

**Comment by samueleruffino99** [comment]
Ok thank you very much!
I will get back to you in case you are interested in also measuring safety/critical scenarios by those metrics.
You do not have any safety metrics in your scenario net tool right?

**Comment by QuanyiLi** [comment]
Indeed, these metrics are missing in both MetaDrive and ScenarioNet but can be useful for training policies, measuring performance, and so on. It is good to have this. Feel free to ask me any questions.

**Comment by samueleruffino99** [comment]
Ok, I will then convert nuScenes (ScenarioNet format) into commonroad format to compute these metrics. However I have some doubts:
- How can I extract lane boarders (left and right), from boundaries in ScenarioNet?
- Which is the coordinate system used in ScenarioNet 
- - They also extract adjacent lines, do you think it would be possible to extract them from scenarionet?

**Comment by QuanyiLi** [comment]
1. Lane borders can be retrieved by `lane.position(longitudinal, lateral)`. Just set later to i.e. 1.5 and iterate from long=0 to long=`lane.length` with a 1-meter interval, and you can extract a lane border. Note: it is likely boarders extracted in this way don't match the solid/dashed lane lines in the dataset. If you want to find a lane line in the dataset as boarder, that could be a bit difficult...
2. Right-handed coordinates
3. Some dataset provides connectivity information, but some not. If the original data provides it, it should be stored in `map_features` as well with keys like `entry_lanes`, `exit_lanes`, `left/right_neighbours`

**Comment by QuanyiLi** [comment]
We are going to directly support OpenDrive in MetaDrive with SUMO's net converter.

**Comment by mimi89999** [comment]
Is there any way in which I could help with development?

**Comment by QuanyiLi** [comment]
@mimi89999 Thank you so much. That feature is almost finished and will be released with the SUMO x MetaDrive bridge soon.

**Comment by sakura-yy520** [comment]
hello，Do you have a tool to import opendrive to metadrive?

**Comment by EllingtonKirby** [comment]
Hello, do you have any updates on this issue? We would be very interested in an opendrive -> scenarionet converter?
